File: backend/app/services/hosted_zone_service.py
from fastapi import HTTPException
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session

from app.models.hosted_zone import HostedZone
from app.schemas.hosted_zone import (
    HostedZoneCreate,
    HostedZoneUpdate,
)


def create_hosted_zone(
    db: Session,
    hosted_zone: HostedZoneCreate,
) -> HostedZone:
    db_hosted_zone = HostedZone(
        name=hosted_zone.name,
        description=hosted_zone.description,
    )

    try:
        db.add(db_hosted_zone)
        db.commit()
        db.refresh(db_hosted_zone)

        return db_hosted_zone

    except IntegrityError:
        db.rollback()

        raise HTTPException(
            status_code=409,
            detail=f"Hosted zone '{hosted_zone.name}' already exists.",
        )


def get_hosted_zones(db: Session) -> list[HostedZone]:
    return db.query(HostedZone).all()


def get_hosted_zone_by_id(
    db: Session,
    hosted_zone_id: int,
) -> HostedZone:
    hosted_zone = (
        db.query(HostedZone)
        .filter(HostedZone.id == hosted_zone_id)
        .first()
    )

    if hosted_zone is None:
        raise HTTPException(
            status_code=404,
            detail="Hosted zone not found.",
        )

    return hosted_zone


def delete_hosted_zone(
    db: Session,
    hosted_zone_id: int,
) -> None:
    hosted_zone = (
        db.query(HostedZone)
        .filter(HostedZone.id == hosted_zone_id)
        .first()
    )

    if hosted_zone is None:
        raise HTTPException(
            status_code=404,
            detail="Hosted zone not found.",
        )

    db.delete(hosted_zone)
    db.commit()

def update_hosted_zone(
    db: Session,
    hosted_zone_id: int,
    hosted_zone_data: HostedZoneUpdate,
) -> HostedZone:
    hosted_zone = (
        db.query(HostedZone)
        .filter(HostedZone.id == hosted_zone_id)
        .first()
    )

    if hosted_zone is None:
        raise HTTPException(
            status_code=404,
            detail="Hosted zone not found.",
        )

    hosted_zone.name = hosted_zone_data.name
    hosted_zone.description = hosted_zone_data.description

    try:
        db.commit()
        db.refresh(hosted_zone)
    except IntegrityError:
        db.rollback()
        raise HTTPException(
            status_code=400,
            detail="Hosted zone already exists.",
        )

    return hosted_zone

File: frontend/features/hosted-zones/services/hostedZoneService.ts
import api from "@/lib/axios";
import type { HostedZone } from "../types.ts";

export interface CreateHostedZoneRequest {
  name: string;
  description: string;
}

export async function getHostedZones(): Promise<HostedZone[]> {
  const response = await api.get("/hosted-zones");
  return response.data;
}

export async function createHostedZone(
  data: CreateHostedZoneRequest
): Promise<HostedZone> {
  const response = await api.post("/hosted-zones", data);
  return response.data;
}
export async function deleteHostedZone(hostedZoneId: number): Promise<void> {
  await api.delete(`/hosted-zones/${hostedZoneId}`);
}
export async function updateHostedZone(
  hostedZoneId: number,
  data: CreateHostedZoneRequest
): Promise<HostedZone> {
  const response = await api.put(`/hosted-zones/${hostedZoneId}`, data);

  return response.data;
}
